### ISSUE TITLE
Epoch Cleaning solution stability

### ISSUE THREAD

**Issue by mdovgialo** [autoreject/autoreject#16]
Hello again,
is it normal for autoreject to have quite different solutions between runs?

Maybe I'm doing something wrong? I'm using ```autoreject.LocalAutoRejectCV(consensus_percs=np.linspace(0, 0.99, 11))
clean_mne = ts.fit_transform(e_mne.pick_types(eeg=True))
```
![solution2](https://cloud.githubusercontent.com/assets/6193380/20637182/c256a022-b37e-11e6-88fa-899dc494c839.png)
![solution1](https://cloud.githubusercontent.com/assets/6193380/20637183/c256f590-b37e-11e6-8833-ca1c3f29dfed.png)

**Comment by jasmainak** [comment]
This is quite possible because it involves cross-validation. It will depend on how you do the splits. You can set the `random_state` like here: https://github.com/autoreject/autoreject/blob/master/examples/plot_visualize_bad_epochs.py#L96. That should hopefully fix it!

**Comment by jasmainak** [comment]
I didn't see your images by the way ... I think you intended to share something

**Comment by mdovgialo** [comment]
That way I'll get the same solution every run, but as you can see one solution might be clearly superior than the other, I guess there is no way in increasing stability? Maybe doing more CV folds and it will eventually find the better solution?

Sorry, somehow if the images are after block of code they don't show up, I fixed it now.

**Comment by jasmainak** [comment]
hmm ... this is weird. I'm guessing you are using the latest code with bayesian optimization? If you can share your data privately, perhaps I can take a look.

**Comment by jasmainak** [comment]
If you want to be adventerous, you could [try changing](https://github.com/autoreject/autoreject/blob/master/autoreject/autoreject.py#L226) `n_calls` here to a higher number to see if it helps ... I haven't exposed this though and I'm not quite certain if it will help

**Comment by mdovgialo** [comment]
Sorry, I'm not sure I'm allowed to share this data. And yes, I was using bayesian_optimisation. I guess it was added later? So random_search is more tested method and it should be more stable? Maybe I should use that?

**Comment by mdovgialo** [comment]
Anyway, thanks for taking your time to explain all this, it will really help. And experimenting with automatic methods always beats manual annotations. So thank you.

**Comment by jasmainak** [comment]
hmm ... I don't think so. Bayesian optimization should work better. It's just a bit slower but it should give more accurate solutions, especially if we increase the number of iterations. I would try the following if you want to dig in further:

- increase `n_calls`
- check the sensor-level thresholds in both cases to see if they are similar

I'm glad someone is using it!

ps: your data looks pretty clean to me. I haven't seen such data in months :)

**Comment by jasmainak** [comment]
If you have any more questions, I will respond tomorrow. It's late in Europe :)

**Comment by mdovgialo** [comment]
You've answered all my questions for now. I know. I wanted to have a little glance at date and that bug, thought "I could ask the original author" and BAM it 3 o'clock in the morning.

Thanks

ps. it's only lucky start of the data, it's quite horrible most of the part.

**Comment by agramfort** [comment]
yes it can happen. I see 2 options

- fix the seed
- use more random splits and draws in bayes opt

if you have such a variability I suspect it's because many thresh are
equally good
so if you run twice you should take the highest estimated threshold to keep
as much
data as possible.

my 2c on a saturday morning...

**Comment by giuliagennari** [comment]
Hi there!
I am facing a very similar issue. 
Running autoreject multiple times on the same subject lead to different outcomes in terms of estimated consensus_percs, number of epochs rejected, "cleanliness" of the signal.
As suggested by Jas, I have tried to fix the seed for the threshold selection and that eventually stabilized the performance (while fixing cv folds did not work for that).

**Comment by guishuyunye-lyw** [comment]
> Hello again,
> is it normal for autoreject to have quite different solutions between runs?
> 
> Maybe I'm doing something wrong? I'm using
> ts = autoreject.LocalAutoRejectCV(consensus_percs=np.linspace(0, 0.99, 11))
> clean_mne = ts.fit_transform(e_mne.pick_types(eeg=True))
> 
> ![solution2](https://cloud.githubusercontent.com/assets/6193380/20637182/c256a022-b37e-11e6-88fa-899dc494c839.png)
> 
> ![solution1](https://cloud.githubusercontent.com/assets/6193380/20637183/c256f590-b37e-11e6-8833-ca1c3f29dfed.png)

how to plot epochs like this!

**Comment by jasmainak** [comment]
yes I would say so. Take a [look at this comment](https://github.com/autoreject/autoreject/issues/16#issuecomment-263053526). The method is a bit sensitive to how deep the cross-validation minima is. Your data looks pretty clean, that's why autoreject finds it difficult to know what is "bad" data. Note that in both runs no epochs are rejected.
